File: FixPoint.fs
module FixPoint 

open Ast

let bottom = fun _ -> None

let id' (x : State) : State option = Some x

let cond (bexpr : State -> bool) (g : State -> State option) (g' : State -> State option) (state : State) : State option = 
    if bexpr state then g state else g' state 

let rec fn (f : (State -> State option) -> (State -> State option)) (n : int) arg =
    match n with 
    | 0 -> arg 
    | _ -> fn f (n - 1) (f arg)

let fix F (state : State) : State = 
    let rec find_fix n state = 
        let lub = fn F n bottom
        match lub state with
        | Some s -> s
        | None   -> find_fix (n+1) state
    find_fix 0 state

File: Haskell/while-interpreter/src/Eval/FixPoint.hs
module Eval.FixPoint (
    fix
    , bottom
    , cond
    , id_p) where
import Parser.Ast ( State )

bottom :: State -> Maybe State
bottom = const Nothing

id_p :: State -> Maybe State
id_p = Just

cond :: (State -> Bool)
        -> (State -> Maybe State)
        -> (State -> Maybe State)
        -> State -> Maybe State
cond bexpr g g' state =
    if bexpr state
        then g state
        else g' state

fn :: ((State -> Maybe State) -> (State -> Maybe State))
      -> Integer -> (State -> Maybe State) -> (State -> Maybe State)
fn _ 0 = id
fn f n = f . fn f (n-1)

fix :: ((State -> Maybe State) -> (State -> Maybe State)) -> State -> State
fix f = lub f 0

lub :: ((State -> Maybe State) -> (State -> Maybe State)) -> Integer -> State -> State
lub f n state = case fn f n bottom state of
    (Just s) -> s
    Nothing -> lub f (n+1) state
